# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/ClientWebGLContext.cpp
# Commit: 8e0aea33f622
# Full Hash: 8e0aea33f622a76c5c8baa7dcf4485ffefd1662f
# Author: Kelsey Gilbert <kelsey.gilbert@mozilla.com>
# Date: 2024-07-09 09:45:28
# Regressor Bug: 1892542
# File Overlap Count: 1
# Description:
#   Bug 1892542 - Webgl e.g. createBuffer() made infallible. r=gfx-reviewers,webidl,saschanaz,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D208050
# ==============================================================================

diff -r b953e0ccb68b -r 8e0aea33f622 dom/canvas/ClientWebGLContext.cpp
--- a/dom/canvas/ClientWebGLContext.cpp	Mon Jul 08 19:36:12 2024 +0300
+++ b/dom/canvas/ClientWebGLContext.cpp	Mon Jul 08 17:24:12 2024 +0000
@@ -941,7 +941,7 @@
   const auto& limits = Limits();
   auto& state = State();
   state.mDefaultTfo = new WebGLTransformFeedbackJS(*this);
-  state.mDefaultVao = new WebGLVertexArrayJS(*this);
+  state.mDefaultVao = new WebGLVertexArrayJS(this);
 
   state.mBoundTfo = state.mDefaultTfo;
   state.mBoundVao = state.mDefaultVao;
@@ -1385,7 +1385,6 @@
 
 already_AddRefed<WebGLBufferJS> ClientWebGLContext::CreateBuffer() const {
   const FuncScope funcScope(*this, "createBuffer");
-  if (IsContextLost()) return nullptr;
 
   auto ret = AsRefPtr(new WebGLBufferJS(*this));
   Run<RPROC(CreateBuffer)>(ret->mId);
@@ -1395,7 +1394,6 @@
 already_AddRefed<WebGLFramebufferJS> ClientWebGLContext::CreateFramebuffer()
     const {
   const FuncScope funcScope(*this, "createFramebuffer");
-  if (IsContextLost()) return nullptr;
 
   auto ret = AsRefPtr(new WebGLFramebufferJS(*this));
   Run<RPROC(CreateFramebuffer)>(ret->mId);
@@ -1406,29 +1404,30 @@
 ClientWebGLContext::CreateOpaqueFramebuffer(
     const webgl::OpaqueFramebufferOptions& options) const {
   const FuncScope funcScope(*this, "createOpaqueFramebuffer");
-  if (IsContextLost()) return nullptr;
 
   auto ret = AsRefPtr(new WebGLFramebufferJS(*this, true));
 
-  const auto& inProcess = mNotLost->inProcess;
-  if (inProcess) {
-    if (!inProcess->CreateOpaqueFramebuffer(ret->mId, options)) {
-      ret = nullptr;
-    }
-    return ret.forget();
-  }
-  const auto& child = mNotLost->outOfProcess;
-  child->FlushPendingCmds();
-  bool ok = false;
-  if (!child->SendCreateOpaqueFramebuffer(ret->mId, options, &ok))
-    return nullptr;
-  if (!ok) return nullptr;
+  if (mNotLost) {
+    const auto& inProcess = mNotLost->inProcess;
+    if (inProcess) {
+      if (!inProcess->CreateOpaqueFramebuffer(ret->mId, options)) {
+        ret = nullptr;
+      }
+      return ret.forget();
+    }
+    const auto& child = mNotLost->outOfProcess;
+    child->FlushPendingCmds();
+    bool ok = false;
+    if (!child->SendCreateOpaqueFramebuffer(ret->mId, options, &ok))
+      return nullptr;
+    if (!ok) return nullptr;
+  }
+
   return ret.forget();
 }
 
 already_AddRefed<WebGLProgramJS> ClientWebGLContext::CreateProgram() const {
   const FuncScope funcScope(*this, "createProgram");
-  if (IsContextLost()) return nullptr;
 
   auto ret = AsRefPtr(new WebGLProgramJS(*this));
   Run<RPROC(CreateProgram)>(ret->mId);
@@ -1437,9 +1436,8 @@
 
 already_AddRefed<WebGLQueryJS> ClientWebGLContext::CreateQuery() const {
   const FuncScope funcScope(*this, "createQuery");
-  if (IsContextLost()) return nullptr;
-
-  auto ret = AsRefPtr(new WebGLQueryJS(*this));
+
+  auto ret = AsRefPtr(new WebGLQueryJS(this));
   Run<RPROC(CreateQuery)>(ret->mId);
   return ret.forget();
 }
@@ -1447,7 +1445,6 @@
 already_AddRefed<WebGLRenderbufferJS> ClientWebGLContext::CreateRenderbuffer()
     const {
   const FuncScope funcScope(*this, "createRenderbuffer");
-  if (IsContextLost()) return nullptr;
 
   auto ret = AsRefPtr(new WebGLRenderbufferJS(*this));
   Run<RPROC(CreateRenderbuffer)>(ret->mId);
@@ -1456,7 +1453,6 @@
 
 already_AddRefed<WebGLSamplerJS> ClientWebGLContext::CreateSampler() const {
   const FuncScope funcScope(*this, "createSampler");
-  if (IsContextLost()) return nullptr;
 
   auto ret = AsRefPtr(new WebGLSamplerJS(*this));
   Run<RPROC(CreateSampler)>(ret->mId);
@@ -1466,7 +1462,6 @@
 already_AddRefed<WebGLShaderJS> ClientWebGLContext::CreateShader(
     const GLenum type) const {
   const FuncScope funcScope(*this, "createShader");
-  if (IsContextLost()) return nullptr;
 
   switch (type) {
     case LOCAL_GL_VERTEX_SHADER:
@@ -1485,7 +1480,6 @@
 already_AddRefed<WebGLSyncJS> ClientWebGLContext::FenceSync(
     const GLenum condition, const GLbitfield flags) const {
   const FuncScope funcScope(*this, "fenceSync");
-  if (IsContextLost()) return nullptr;
 
   if (condition != LOCAL_GL_SYNC_GPU_COMMANDS_COMPLETE) {
     EnqueueError_ArgEnum("condition", condition);
@@ -1511,7 +1505,6 @@
 
 already_AddRefed<WebGLTextureJS> ClientWebGLContext::CreateTexture() const {
   const FuncScope funcScope(*this, "createTexture");
-  if (IsContextLost()) return nullptr;
 
   auto ret = AsRefPtr(new WebGLTextureJS(*this));
   Run<RPROC(CreateTexture)>(ret->mId);
@@ -1521,7 +1514,6 @@
 already_AddRefed<WebGLTransformFeedbackJS>
 ClientWebGLContext::CreateTransformFeedback() const {
   const FuncScope funcScope(*this, "createTransformFeedback");
-  if (IsContextLost()) return nullptr;
 
   auto ret = AsRefPtr(new WebGLTransformFeedbackJS(*this));
   Run<RPROC(CreateTransformFeedback)>(ret->mId);
@@ -1531,9 +1523,8 @@
 already_AddRefed<WebGLVertexArrayJS> ClientWebGLContext::CreateVertexArray()
     const {
   const FuncScope funcScope(*this, "createVertexArray");
-  if (IsContextLost()) return nullptr;
-
-  auto ret = AsRefPtr(new WebGLVertexArrayJS(*this));
+
+  auto ret = AsRefPtr(new WebGLVertexArrayJS(this));
   Run<RPROC(CreateVertexArray)>(ret->mId);
   return ret.forget();
 }
@@ -6678,14 +6669,15 @@
 
 // ---------------------------
 
-webgl::ObjectJS::ObjectJS(const ClientWebGLContext& webgl)
-    : mGeneration(webgl.mNotLost), mId(webgl.NextId()) {}
+webgl::ObjectJS::ObjectJS(const ClientWebGLContext* const webgl)
+    : mGeneration(webgl ? webgl->mNotLost : nullptr),
+      mId(webgl ? webgl->NextId() : 0) {}
 
 // -
 
 WebGLFramebufferJS::WebGLFramebufferJS(const ClientWebGLContext& webgl,
                                        bool opaque)
-    : webgl::ObjectJS(webgl), mOpaque(opaque) {
+    : webgl::ObjectJS(&webgl), mOpaque(opaque) {
   (void)mAttachments[LOCAL_GL_DEPTH_ATTACHMENT];
   (void)mAttachments[LOCAL_GL_STENCIL_ATTACHMENT];
   if (!webgl.mIsWebGL2) {
@@ -6697,6 +6689,8 @@
 
 void WebGLFramebufferJS::EnsureColorAttachments() {
   const auto& webgl = Context();
+  if (!webgl) return;  // Context is lost.
+
   const auto& limits = webgl->Limits();
   auto maxColorDrawBuffers = limits.maxColorDrawBuffers;
   if (!webgl->mIsWebGL2 &&
@@ -6709,7 +6703,7 @@
 }
 
 WebGLProgramJS::WebGLProgramJS(const ClientWebGLContext& webgl)
-    : webgl::ObjectJS(webgl),
+    : webgl::ObjectJS(&webgl),
       mKeepAlive(std::make_shared<webgl::ProgramKeepAlive>(*this)),
       mKeepAliveWeak(mKeepAlive) {
   (void)mNextLink_Shaders[LOCAL_GL_VERTEX_SHADER];
@@ -6719,18 +6713,19 @@
 }
 
 WebGLShaderJS::WebGLShaderJS(const ClientWebGLContext& webgl, const GLenum type)
-    : webgl::ObjectJS(webgl),
+    : webgl::ObjectJS(&webgl),
       mType(type),
       mKeepAlive(std::make_shared<webgl::ShaderKeepAlive>(*this)),
       mKeepAliveWeak(mKeepAlive) {}
 
 WebGLTransformFeedbackJS::WebGLTransformFeedbackJS(
     const ClientWebGLContext& webgl)
-    : webgl::ObjectJS(webgl),
+    : webgl::ObjectJS(&webgl),
       mAttribBuffers(webgl::kMaxTransformFeedbackSeparateAttribs) {}
 
-WebGLVertexArrayJS::WebGLVertexArrayJS(const ClientWebGLContext& webgl)
-    : webgl::ObjectJS(webgl), mAttribBuffers(webgl.Limits().maxVertexAttribs) {}
+WebGLVertexArrayJS::WebGLVertexArrayJS(const ClientWebGLContext* const webgl)
+    : webgl::ObjectJS(webgl),
+      mAttribBuffers(webgl ? webgl->Limits().maxVertexAttribs : 0) {}
 
 // -
 